association_backend/association/api/tests/mutations/test_do_checkout.py

```diff
@@ -0,0 +1,105 @@
+import datetime
+from unittest.mock import patch
+from zoneinfo import ZoneInfo
+
+from association.api.tests.graphql_client import graphql_client
+from association.domain.exceptions import AlreadySubscribed
+from association.tests.factories import SubscriptionFactory
+from association.tests.session import db
+from ward import skip, test
+
+rome_tz = ZoneInfo("Europe/Rome")
+
+
+@test("ok")
+async def _(graphql_client=graphql_client, db=db):
+    query = """
+    mutation {
+        doCheckout {
+            __typename
+            ... on SubscriptionResponse {
+                __typename
+                subscription {
+                    creationDate
+                    stripeCustomerId
+                    userId
+                    state
+                    stripeSessionId
+                    expirationDate
+                    paymentDate
+                    stripeId
+                }
+            }
+        }
+    }
+    """
+    with patch(
+        "association.domain.services.do_checkout",
+        return_value=SubscriptionFactory(stripe_session_id="cs_test_12345"),
+    ) as do_checkout_mock:
+        response = await graphql_client.query(query, variables={})
+        do_checkout_mock.assert_called_once()
+        assert response.data["doCheckout"]["__typename"] == "SubscriptionResponse"
+        assert (
+            response.data["doCheckout"]["subscription"]["stripeSessionId"]
+            == "cs_test_12345"
+        )
+
+
+@test("Already Subscribed")
+async def _(graphql_client=graphql_client, db=db):
+    query = """
+    mutation {
+        doCheckout {
+            __typename
+            ... on AlreadySubscribedError {
+                __typename
+                expirationDate
+                message
+            }
+        }
+    }
+    """
+    mocked_datetime = datetime.datetime.now(rome_tz) - datetime.timedelta(days=1 + 365)
+    with patch("association.domain.services.do_checkout") as do_checkout_mock:
+        do_checkout_mock.side_effect = AlreadySubscribed(
+            expiration_date=mocked_datetime
+        )
```

Comment:
uhm, we should try mocking only the stripe calls

also, I wonder if there is a better way to mock Stripe

Reply:
the test you want to see is already covered in `tests/services/test_do_checkout.py`
Here I need isolation to be sure that I return the right response when AlreadySubscribed is raised, this is not a E2E test.
Do you want  E2E tests in tests/mutations/ ?
Elsewhere i can create a dir tests/e2e/ to cover all the flows entirely.
--------------
I wonder too, and I have not found anything decent honestly.
I expected stripe client to provide factories, I'll make something up